### ISSUE TITLE
Sporadically older time series are not returned

### ISSUE THREAD

**Issue by mpoindexter** [VictoriaMetrics/VictoriaMetrics#225]
**Describe the bug**
Sometimes data for older time series are not returned in a query that includes that date range. For example, if I have `my_metric{instance="a"}` that covers the time period 2019-10-01 10:00 AM to 2019-10-01 11:00 AM and `my_metric{instance="b"}` that covers the time period 2019-10-01 11:00 AM to 2019-10-02 12:00 AM, sometimes if I issue a query for `my_metric` for date range 2019-10-01 10:00 AM to 2019-10-02 12:00 AM only `my_metric{instance="b"}` is returned. After some time, the situation corrects, and both time series are returned. After a reboot, the problem goes away, then returns. No errors appear in the log.

**To Reproduce**
Unclear

**Expected behavior**
All time series matching the query/time range are returned

**Screenshots**

**Version**
1.28.2

**Additional context**
I think the problem is here: https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/lib/storage/index_db.go#L965

This method caches things in a tag cache by a key that includes only tag filters, but then populates the cache of time series matching via logic that takes into account the time range of the query. The actual logic is rather complicated, and only sometimes takes into account the time range, so I've not been able to determine precisely what conditions trigger it, but I think it's incorrect to store things in the tagCache when they can only be valid for a certain TimeRange. If my theory is correct for this being the cause, it would match the behavior I see since new generations of the tag filters key would cause the situation to automatically correct over time as long as a query with a long enough time range arrives first after a reset.

**Comment by valyala** [comment]
Thanks for the detailed bug report! It is probably fixed yesterday in c51ca04a4340a9f01fcbe82ddb7cb6aea9d32b11 . Could you build VictoriaMetrics from this commit and verify whether the issue is fixed?

**Comment by mpoindexter** [comment]
I was able to write a unit test that reproduced the problem, looks like c51ca04 fixes it. Here's the test if you think it's useful to add to the test suite.

```
func TestSearchTSIDWithTimeRange(t *testing.T) {
	metricIDCache := workingsetcache.New(1234, time.Hour)
	metricNameCache := workingsetcache.New(1234, time.Hour)
	defer metricIDCache.Stop()
	defer metricNameCache.Stop()

	currMetricIDs := &hourMetricIDs{
		isFull: true,
		m:      &uint64set.Set{},
	}

	var hmCurr atomic.Value
	hmCurr.Store(currMetricIDs)

	prevMetricIDs := &hourMetricIDs{
		isFull: true,
		m:      &uint64set.Set{},
	}
	var hmPrev atomic.Value
	hmPrev.Store(prevMetricIDs)

	dbName := "test-index-db-ts-range"
	db, err := openIndexDB(dbName, metricIDCache, metricNameCache, &hmCurr, &hmPrev)
	if err != nil {
		t.Fatalf("cannot open indexDB: %s", err)
	}
	defer func() {
		db.MustClose()
		if err := os.RemoveAll(dbName); err != nil {
			t.Fatalf("cannot remove indexDB: %s", err)
		}
	}()

	is := db.getIndexSearch()
	defer db.putIndexSearch(is)

	// Create a bunch of per-day time series
	const days = 5
	const metricsPerDay = 1000
	theDay := time.Date(2019, time.October, 15, 5, 1, 0, 0, time.UTC)
	now := uint64(timestampFromTime(theDay))
	currMetricIDs.hour = now / msecPerHour
	prevMetricIDs.hour = (now - msecPerHour) / msecPerHour
	baseDate := now / msecPerDay
	var metricNameBuf []byte
	for day := 0; day < days; day++ {
		var tsids []TSID
		for metric := 0; metric < metricsPerDay; metric++ {
			var mn MetricName
			mn.MetricGroup = []byte("testMetric")
			mn.AddTag(
				"constant",
				"const",
			)
			mn.AddTag(
				"day",
				fmt.Sprintf("%v", day),
			)
			mn.AddTag(
				"uniqueid",
				fmt.Sprintf("%v", metric),
			)
			mn.sortTags()

			metricNameBuf = mn.Marshal(metricNameBuf[:0])
			var tsid TSID
			if err := is.GetOrCreateTSIDByName(&tsid, metricNameBuf); err != nil {
				t.Fatal(fmt.Errorf("unexpected error when creating tsid for mn:\n%s: %s", &mn, err))
				return
			}
			tsids = append(tsids, tsid)
		}

		// Add the metrics to the per-day stores
		date := baseDate - uint64(day*msecPerDay)
		for i := range tsids {
			tsid := &tsids[i]
			if err := db.storeDateMetricID(date, tsid.MetricID); err != nil {
				t.Fatal(fmt.Errorf("error in storeDateMetricID(%d, %d): %s", date, tsid.MetricID, err))
				return
			}
		}

		// Add the the hour metrics caches
		if day == 0 {
			for i := 0; i < 256; i++ {
				prevMetricIDs.m.Add(tsids[i].MetricID)
				currMetricIDs.m.Add(tsids[i].MetricID)
			}
		}
	}

	// Flush index to disk, so it becomes visible for search
	db.tb.DebugFlush()

	// Create a filter that will match series that occur across multiple days
	var tfs TagFilters
	tfs.Reset()
	if err := tfs.Add([]byte("constant"), []byte("const"), false, false); err != nil {
		t.Fatalf("cannot add filter: %s", err)
	}

	// Perform a search that can be fulfilled out of the hour metrics cache.
	// This should return the metrics in the hourly cache
	tr := TimeRange{
		MinTimestamp: int64(now - msecPerHour - 59999),
		MaxTimestamp: int64(now),
	}
	matchedTSIDs, err := db.searchTSIDs([]*TagFilters{&tfs}, tr, 10000)
	if len(matchedTSIDs) != 256 {
		t.Fatal("Expected time series for current hour, got", len(matchedTSIDs))
	}

	// Perform a search within a day that falls out out of the hour metrics cache.
	// This should return the metrics for the day
	tr = TimeRange{
		MinTimestamp: int64(now - msecPerHour - 60001),
		MaxTimestamp: int64(now),
	}
	matchedTSIDs, err = db.searchTSIDs([]*TagFilters{&tfs}, tr, 10000)
	if len(matchedTSIDs) != metricsPerDay {
		t.Fatal("Expected time series for current day, got", len(matchedTSIDs))
	}

	// Perform a search across all the days, should match all metrics
	tr = TimeRange{
		MinTimestamp: int64(now - msecPerDay*days),
		MaxTimestamp: int64(now),
	}

	matchedTSIDs, err = db.searchTSIDs([]*TagFilters{&tfs}, tr, 10000)
	if len(matchedTSIDs) != metricsPerDay*days {
		t.Fatal("Expected time series for all days, got", len(matchedTSIDs))
	}
}
```

**Comment by valyala** [comment]
@mpoindexter , it would be great if you could send a pull request with this unit test, so I could merge it.

**Comment by mpoindexter** [comment]
I'd be happy to. I wasn't sure if this unit test was quite correct or the best way to structure things, but if it seems right I will do a PR.

**Comment by valyala** [comment]
Yes, just send a PR with the test for this issue!

FYI, the bugfix is available in the [v1.29.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases).

**Comment by valyala** [comment]
I'm closing the issue as fixed, but still waiting for the PR with the test.
